sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -446,7 +451,12 @@ case class DataSource(
         dataSource.createRelation(
           sparkSession.sqlContext, mode, caseInsensitiveOptions, Dataset.ofRows(sparkSession, data))
       case format: FileFormat =>
-        sparkSession.sessionState.executePlan(planForWritingFileFormat(format, mode, data)).toRdd
+        val qe = sparkSession.sessionState.executePlan(planForWritingFileFormat(format, mode, data))
+        qe.executedPlan.transform {
+          case f: FileWritingCommandExec =>
+            val newCmd = f.cmd.withExternalMetrics(externalMetrics.getOrElse(null))
+            FileWritingCommandExec(newCmd, f.children)
+        }.execute()
```

Comment:
Hm..

So it looks like you do this in 2 places: `CreateHiveTableAsSelectCommand` and `CreateDataSourceTableAsSelectCommand.saveDataIntoTable() -> DataSource.writeAndRead()`

For the latter case, I'd pass `CreateDataSourceTableAsSelectCommand.metrics` through `planForWritingFileFormat()` to `InsertIntoHadoopFsRelationCommand` which already extends `FileWritingCommand` and could just take those existing metrics as a constructor argument.

For the former, I'm not sure, as I see `InsertIntoTable` does not extend `FileWritingCommand`. Maybe it should? Then a similar solution would apply there.